### PR TITLE
Lets Encrypt Integration

### PR DESCRIPTION
**Pull request by pzmrzy**
When first time run the server, it will automatic get certificate from Lets Encrypt
For staging environment:
`sudo ./bin/dispatch-server-linux --host 0.0.0.0 --enable-tls --lets-encrypt --domain <domain>`
For production environment:
`sudo ./bin/dispatch-server-linux --host 0.0.0.0 --enable-tls --lets-encrypt --domain <domain> --production`
#668

### CODE REVIEW

File: pkg/dispatchserver/config.go [hunk 1]
```diff
@@ -36,8 +36,11 @@ type serverConfig struct {
 	DisableHTTP       bool   `mapstructure:"disable-http" json:"disable-http"`
 	TLSPort           int    `mapstructure:"tls-port" json:"tls-port"`
 	EnableTLS         bool   `mapstructure:"enable-tls" json:"enable-tls"`
+	LetsEncrypt       bool   `mapstructure:"lets-encrypt" json:"lets-encrypt"`
+	Production        bool   `mapstructure:"production" json:"production"`
```
Comment (kars7e) [inline pkg/dispatchserver/config.go]: IMHO you can drop this flag. staging environment is used for testing, e.g. when you are developing automation for that. you could have an internal constant, or a config flag to set the let's encrypt server, but production server should be the default.

File: pkg/dispatchserver/config.go [hunk 2]
```diff
@@ -36,8 +36,11 @@ type serverConfig struct {
...
 	TLSCertificate    string `mapstructure:"tls-certificate" json:"tls-certificate"`
 	TLSCertificateKey string `mapstructure:"tls-certificate-key" json:"tls-certificate-key"`
+	Domain            string `mapstructure:"domain" json:"domain"`
```
Comment (kars7e) [inline pkg/dispatchserver/config.go]: If the domain is needed, how about we merge lets-encrypt and this flag and have only one, `lets-encrypt-domain` flag? If it's empty, we disable lets encrypt. if it's set, we use the value as the domain.

File: pkg/utils/constants.go [hunk 3]
```diff
@@ -52,3 +52,18 @@ const ServiceAccountKind = "ServiceAccount"
 
 // OrganizationKind a constant representing the kind of the Organization Model
 const OrganizationKind = "Organization"
+
+// LetsEncryptStaging a constant representing Let's Encrypt staging server url
+const LetsEncryptStaging = "https://acme-staging.api.letsencrypt.org/directory"
```
Comment (kars7e) [inline pkg/utils/constants.go]: these constants look http-specific, move them to `http` package. In general, avoid `utils` package if possible.

File: pkg/http/server.go [hunk 4]
```diff
@@ -185,6 +199,24 @@ func (s *Server) Serve() (err error) {
...
+		if s.LetsEncrypt {
+			if s.Domain == "" {
+				return errors.New("Need a domain name, use --domain <Domain Name>")
+			}
+			if !s.Production {
+				log.Warnf("Staging environment, use --production for production environment")
+			}
+			uDomain := strings.Replace(s.Domain, ".", "_", -1)
+			if s.existCertificate(uDomain) {
+				s.TLSCertificate = "./lets_encrypt/" + uDomain + ".crt"
```
Comment (kars7e) [inline pkg/http/server.go]: `./lets_encrypt/` should be a constant, you are using it in multiple places.

File: pkg/http/server.go [hunk 5]
```diff
@@ -381,3 +413,164 @@ func (s *Server) HTTPSURL() string {
...
+
+	// Attempt authorization
+	auth, err := c.Authorize(context.TODO(), s.Domain)
+	if err != nil {
+		return err
+	}
+	var challenge *acme.Challenge
+	for _, c := range auth.Challenges {
+		if c.Type == "http-01" {
+			challenge = c
```
Comment (kars7e) [inline pkg/http/server.go]: I'm guessing you are interested in the first occurrence of a challenge of type `http-01`, if so you could add `break` here to break out of the loop as soon as you find it.

File: pkg/http/server.go [hunk 6]
```diff
@@ -381,3 +413,164 @@ func (s *Server) HTTPSURL() string {
...
+	log.Infof("Registration succeeded")
+
+	// Attempt authorization
+	auth, err := c.Authorize(context.TODO(), s.Domain)
+	if err != nil {
+		return err
+	}
+	var challenge *acme.Challenge
+	for _, c := range auth.Challenges {
+		if c.Type == "http-01" {
```
Comment (kars7e) [inline pkg/http/server.go]: http-01 means a chalange over unauthenticated HTTP and port 80, which means people who want to listen only on port 443 will have to open one more port in the firewall. let's talk more about this offline.

File: pkg/http/server.go [hunk 7]
```diff
@@ -381,3 +413,164 @@ func (s *Server) HTTPSURL() string {
...
+	if _, err := os.Stat("./lets_encrypt/" + uDomain + ".key"); os.IsNotExist(err) {
+		return false
+	}
+	if _, err := os.Stat("./lets_encrypt/" + uDomain + ".crt"); os.IsNotExist(err) {
+		return false
+	}
+	return true
+}
+
+func (s *Server) getLetsEncrypt(uDomain string) error {
```
Comment (kars7e) [inline pkg/http/server.go]: this function is really long, let's break it down into smaller pieces. take a look at `crypto.go`, maybe we can share some logic with it.

File: pkg/http/server.go [hunk 8]
```diff
@@ -381,3 +413,164 @@ func (s *Server) HTTPSURL() string {
...
+	}
+
+	log.Infof("Complete!")
+
+	s.TLSCertificateKey = fmt.Sprintf("%s.key", path.Join(certDirectory, uDomain))
+	s.TLSCertificate = path.Join(certDirectory, fmt.Sprintf("%s.crt", uDomain))
+	return nil
+}
+
+func newKey(filename string) (crypto.Signer, error) {
```
Comment (kars7e) [inline pkg/http/server.go]: `crypto.go` already generates private key, maybe we could reuse that (also it uses ecdsa, which has the advantage of a smaller key size).

File: pkg/utils/constants.go [hunk 9]
```diff
@@ -52,3 +52,18 @@ const ServiceAccountKind = "ServiceAccount"
...
+
+// LetsEncryptProduction a constant representing Let's Encrypt production server url
+const LetsEncryptProduction = "https://acme-v01.api.letsencrypt.org/directory"
+
+// KeyLength a constant representing the length of TLS Key
+const KeyLength = 2048
+
+// KeyType a constant representing the length of TLS Key Type
```
Comment (tenczar) [inline pkg/utils/constants.go]: Comment is a copy of the previous comment.

File: pkg/http/server.go (hunk 4 above)
Comment (tenczar) [reply pkg/http/server.go]: I would go a step further and say this should be a configurable parameter. If we are going to be writing files we should give the end user a choice of where those files go.

File: pkg/http/server.go [hunk 10]
```diff
@@ -381,3 +413,164 @@ func (s *Server) HTTPSURL() string {
...
+	}
+	if _, err := os.Stat("./lets_encrypt/" + uDomain + ".crt"); os.IsNotExist(err) {
+		return false
+	}
+	return true
+}
+
+func (s *Server) getLetsEncrypt(uDomain string) error {
+	letsEncryptURL := utils.LetsEncryptStaging
+	certDirectory := "./lets_encrypt"
```
Comment (tenczar) [inline pkg/http/server.go]: I would move this below the following `if` block. Put all of the initialization of `letsEncryptURL` in one place.

File: pkg/http/server.go [hunk 11]
```diff
@@ -381,3 +413,164 @@ func (s *Server) HTTPSURL() string {
...
+
+	// Create the ACME client
+	c := &acme.Client{
+		Key:          aKey,
+		DirectoryURL: letsEncryptURL,
+	}
+
+	// Begin with registration
+	log.Infof("Attempting registration...")
+	_, err = c.Register(context.TODO(), nil, func(string) bool { return true })
```
Comment (tenczar) [inline pkg/http/server.go]: you can do `_, err := c.Register(context.TODO(), nil, acme.AcceptTOS)` to make this a little more clear to a human reader.

File: pkg/http/server.go [hunk 12]
```diff
@@ -381,3 +413,164 @@ func (s *Server) HTTPSURL() string {
 	}
 	return ""
 }
+
+func (s *Server) existCertificate(uDomain string) bool {
```
Comment (berndtj) [inline pkg/http/server.go]: existingCertificate

**Pull request closed without merging**